runtime/mux.go

```diff
@@ -239,34 +226,22 @@ func (s *ServeMux) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			// X-HTTP-Method-Override is optional. Always allow fallback to POST.
 			if s.isPathLengthFallback(r) {
 				if err := r.ParseForm(); err != nil {
-					if s.protoErrorHandler != nil {
-						_, outboundMarshaler := MarshalerForRequest(s, r)
-						sterr := status.Error(codes.InvalidArgument, err.Error())
-						s.protoErrorHandler(ctx, s, outboundMarshaler, w, r, sterr)
-					} else {
-						OtherErrorHandler(w, r, err.Error(), http.StatusBadRequest)
-					}
+					_, outboundMarshaler := MarshalerForRequest(s, r)
+					sterr := status.Error(codes.InvalidArgument, err.Error())
+					s.errorHandler(ctx, s, outboundMarshaler, w, r, sterr)
 					return
 				}
 				h.h(w, r, pathParams)
 				return
 			}
-			if s.protoErrorHandler != nil {
-				_, outboundMarshaler := MarshalerForRequest(s, r)
-				s.protoErrorHandler(ctx, s, outboundMarshaler, w, r, ErrUnknownURI)
-			} else {
-				OtherErrorHandler(w, r, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
-			}
+			_, outboundMarshaler := MarshalerForRequest(s, r)
+			s.errorHandler(ctx, s, outboundMarshaler, w, r, ErrUnknownURI)
 			return
 		}
 	}
 
-	if s.protoErrorHandler != nil {
-		_, outboundMarshaler := MarshalerForRequest(s, r)
-		s.protoErrorHandler(ctx, s, outboundMarshaler, w, r, ErrUnknownURI)
-	} else {
-		OtherErrorHandler(w, r, http.StatusText(http.StatusNotFound), http.StatusNotFound)
-	}
+	_, outboundMarshaler := MarshalerForRequest(s, r)
+	s.errorHandler(ctx, s, outboundMarshaler, w, r, ErrUnknownURI)
```

Comment:
The handler now returns `http.StatusNotImplemented` rather than `http.StatusNotFound` when no path was found. This is more consistent with gRPC in general.

Reply:
> This is more consistent with gRPC in general.

But it is totally inconsistent with typical REST APIs. The HTTP client for grpc-gateway is using a REST API so those idioms feel much more appropriate than trying to mirror gRPC's behavior.

Reply:
I'm happy to reverse this change, it wasn't clear to me which way to go. The same argument can be extended to error formats in general, do you think it makes sense to rely on the status.Status type as an error still?

Reply:
I think using `status.Status` for every error is nice for uniformity. But since it's was designed to convey the result of an RPC operation, it can be awkward for certain kinds of errors.

Perhaps when this status is constructed, you could add a gateway-specific error detail to it, making it clear that the error was caused by not being able to map the given URI to a known RPC method. Then, as long as the hook for customizing the error response is used for these errors too, you could do the "not implemented" thing by default. Those that prefer to render the errors as "not found" can still do so with a custom error handler while also being able to preserve "not implemented" errors from other situations (like when that is actually the code in the RPC server's response).

Reply:
I think you're right that this error code is going to be confusing for most users, and I'm simply going to revert the change in behaviour. Thanks for your feedback!